### ISSUE TITLE
ndarray of Vectors with SOA layout crashes

### ISSUE THREAD

**Issue by RuRo** [taichi-dev/taichi#4566]
Consider the following:

```python
import taichi as ti
import numpy as np

ti.init()

def check_soa(n, m):
    a = ti.Vector.ndarray(n, ti.i32, (m,), ti.Layout.SOA)
    b = 1 + np.arange(n * m).reshape(n, m).astype(np.int32)
    a.from_numpy(b)

    print("Expected:")
    print(b.flatten())

    print("Got:")
    print(a.to_numpy().flatten())
    print()


check_soa(10, 2)
check_soa(2, 10)
```

Results in
```
Expected:
[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]
Got:
[ 1  2  3  4  5  6  7  8  9 10 11 12  0  0  0  0  0  0  0  0]

Expected:
[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]
Got:
[ 1  2  0  0  0  0  0  0  0  0 11 12  0  0  0  0  0  0  0  0]
[E 03/18/22 00:13:50.660 639004] Received signal 11 (Segmentation fault)
```

Bonus (probably unrelated) issue:
```python
import taichi as ti
# note that there is no ti.init()
ti.ndarray(ti.f32, (1, 1))
# segmentation fault (core dumped)
```

**Comment by RuRo** [comment]
Here's one more segfault. This one is due to the `offset` parameter:

```python
import taichi as ti
import numpy as np

ti.init()

def check_off(n, off):
    a = ti.field(ti.f32, (n,), offset=off)
    a.fill(1)

    print("Expected:")
    print(np.ones(n).flatten())

    print("Got:")
    print(a.to_numpy().flatten())
    print()


check_off(16, 8)
check_off(16, -8)
```

```
Expected:
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
Got:
[0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1.]

Expected:
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
Got:
[1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
[E 03/18/22 02:51:17.184 899651] Received signal 11 (Segmentation fault)
```

**Comment by strongoier** [comment]
Regarding the problem of ndarray `to_numpy` cc: @qiao-bo

**Comment by strongoier** [comment]
I will close this issue for now because:
- the first problem is solved by #4597
- the bonus issue is the same kind of problem as #4471
- the other segfault problem duplicates #1596.
